### ISSUE TITLE
[1.8.1 RHEL 6.x] Service shows incorrectly as running locally, but stopped remotely, regardless of actual status

### ISSUE THREAD

**Issue by jm-welch** [NagiosEnterprises/ncpa#242]
Stood up a new service (`corestreet-va`) on RHEL 6.8, and trying to monitor it via NCPA.

``` shell
[root@dc3qszva01 ~]# service corestreet-va status
corestreet-va is not running

[root@dc3qszva01 ~]# echo $?
1

[root@dc3qszva01 ~]# curl --insecure -s --data "token=mbRnmyct" 'https://localhost:5693/api/services'
{
  "value": {
    "services": {
      # snipped full service list
      "ncpa_listener": "stopped",
      "ncpa_passive": "stopped",
      "corestreet-va": "stopped"
    }
  }
}

[root@dc3qszva01 ~]# curl --insecure -s --data "token=mbRnmyct" 'https://localhost:5693/api/service/corestreet-va'
{
  "value": {
    "services": {
      "corestreet-va": "running"
    }
  }
}
```

Tested the code from [services.py](https://github.com/NagiosEnterprises/ncpa/blob/master/agent/listener/services.py#L152-L180), since `get_services_via_initd` is called both as the fallback and as part of `get_services_via_initctl`. In the latter, initctl is used to confirm the results from the initd function, but the `corestreet-va` service is not present in `initctl list`, so it would rely on the status determined during `get_services_via_initd`.

``` python
>>> from stat import ST_MODE,S_IXUSR,S_IXGRP,S_IXOTH
>>> import os, re, subprocess, platform, tempfile
>>> possible_services = filter(lambda x: os.stat('/etc/init.d/'+x)[ST_MODE] & (S_IXUSR|S_IXGRP|S_IXOTH), os.listdir('/etc/init.d'))
>>> possible_services
['ip6tables', 'spice-vdagentd', 'rdisc', 'crond', 'wdaemon', 'sssd', 'ncpa_listener', 'rpcbind', 'killall', 'autofs', 'sshd', 'netconsole', 'ntpdate', 'corestreet-va', 'lvm2-monitor', 'udev-post', 'auditd', 'ntpd', 'rhnsd', 'rpcsvcgssd', 'jexec', 'nfs', 'rpcgssd', 'oddjobd', 'saslauthd', 'firstboot', 'rpcidmapd', 'network', 'messagebus', 'certmonger', 'postfix', 'iptables', 'nfslock', 'postgresql-9.4', 'lvm2-lvmetad', 'restorecond', 'blk-availability', 'netfs', 'rhsmcertd', 'ncpa_passive', 'single', 'halt', 'sandbox', 'rsyslog', 'haldaemon']
>>> grep_search = '[%s]%s' % (possible_services[13][0], possible_services[13][1:])
>>> grep_search
'[c]orestreet-va'
>>> ps_call = subprocess.Popen(['ps', 'aux'], stdout=subprocess.PIPE)
>>> grep_call = subprocess.Popen(['grep', grep_search], stdout=subprocess.PIPE, stdin=ps_call.stdout)
>>> ps_call.wait()
0
>>> ps_call.stdout.close()
>>> grep_call.wait()
1
>>> grep_call.returncode
1
```

Since [the next lines](https://github.com/NagiosEnterprises/ncpa/blob/master/agent/listener/services.py#L176-L177) set the status to running only if that `grep_call.returncode` is `0`, and the [default setting](https://github.com/NagiosEnterprises/ncpa/blob/master/agent/listener/services.py#L156) for all services in the list is `stopped`, this should return `stopped` for this service.

``` python
if grep_call.returncode == 0:
    services[service] = 'running'
```

This looks similar to #142, in that a service's status is always reported in a given state, regardless of its actual state. Also similar to #137, which discusses the issue on linux NCPA listeners when a service's name and its binary aren't the same (as in this case - the actual service running is java, as this is a tomcat service). What's odd here is that it's always reporting `running` when executed locally on the host, but always reporting `stopped` when executed remotely.

If I can provide any other information, or help build a test case to confirm that this isn't present in the 2.0 build, please let me know - I'd be glad to assist.

**Comment by jomann09** [comment]
When you were testing this did you test it using the latest version of the python code from GitHub? I actually changed some of the way it checked because of some of the strange logic used to determine what was considered running and not, including backup checks to make sure things were actually running or not.

**Comment by jm-welch** [comment]
I used the code from the page I linked, which is from master. I didn't check any development branches yet.

While I know that not every service does this, it might be worthwhile to check the exit code of the `service x status` command, too. If I write an init script, I make sure that it exits non-0 if the service isn't running.

At this point, though, it sounds like you'd have to build a ton of conditions:

``` python
for service in servicelist:
    initstatus = getServiceInitStatus(service) # define this magic elsewhere
    initctlstatus = getServiceInitCtlStatus(service) # this too
    if any(status.exitcode != 0, any(st in status.stdout.lower() for st in ['not running', 'stopped']), 'stopped' in initctlstatus.stdout.lower(), #blah, #blah):
        services[service] = 'stopped'
```

I'm sure there's a more elegant way, but that's what I came up with while waiting for an ansible play to run. I'll keep digging tomorrow, and maybe do some test builds to see what I can find.

**Comment by jomann09** [comment]
I actually don't disagree with checking the status code but we may have to do it much like how you showed it above. Sadly that might just be the most effective way. I think that a combination of checking like that and how it currently checks with ps aux, and making the way you did above be the overriding one we could definitely catch most cases. Other systems have a lot better ways to list services and show more information in a very generic way so it's mostly just initd systems we have to worry about.

**Comment by jm-welch** [comment]
To confirm what I'm seeing on the network graph - all current code is in master, right? No development branches/forks I should be aware of at the moment?

**Comment by jomann09** [comment]
Yup! Once 2.0.0 is finished, I will start working on branches instead of master.

I was doing a cleanup of past branches by Nick since some of them overlapped with functionality. I am going to, after the 2.0.0 version, add back in continuous integration/testing but currently the testing is in shambles. I think we should be able to use Travis CI and AppVeyor and put the tests somewhere that is out of the way of the actual project source code.

**Comment by jm-welch** [comment]
Last question, somewhat unrelated - has anyone put together a vagrantfile to build a build VM? If not, I'll start work on one, see what I can pull together.

**Comment by jomann09** [comment]
I believe @lgroschen worked on this at one time, and it is possible to re-add it to the project if it gets some testing to make sure it works.

**Comment by jomann09** [comment]
Sorry I haven't been around much, was sick over the holiday weekend and have been busy at work! I am going to take a look at some of this stuff tonight and hopefully get some things sorted out! I'll get some new builds up sometime tomorrow also for testing.

**Comment by jomann09** [comment]
Working on this right now. I don't have your example, but looking through some services I foundone already that isn't working. `haldaemon`  shows as stopped and causes the same issues. I'll use it to test the solution we talked about above to make it accurate.

**Comment by jomann09** [comment]
Closing this for now since it is fixed with 8a3502e731e8615522caeb7077d4995fe2ec31c7

I realize this is a pretty gross way of doing it. I made a huge check that basically does a lot of assumptions if it can't figure out if it's running or not based on the return code, output or no output, or if the output shows not running etc.

Unfortunately there probably isn't a good way to catch everything, but looking through all of the cases I found I see they are all showing the proper status. Probably something that should be tested just to verify.

As a side note - I did have it running the service checks first before checking via `ps aux` because I wanted a better check to start with but it seemed to take about 30% longer to grab the entire `api/services` list like that so I am now doing it after the `ps aux` check.
